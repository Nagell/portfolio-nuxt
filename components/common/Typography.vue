<template>
    <component
        :is="tag"
        :class="cn(variantClasses, props.class)"
    >
        <slot />
    </component>
</template>

<script lang="ts" setup>
    import { cn } from '~/lib/utils'

    import type { HTMLAttributes } from 'vue'

    type Variant = 'h1' | 'h2' | 'h3' | 'h4' | 'p' | 'ul' | 'li' | 'code'

    interface Props {
        tag?: Variant
        variant?: Variant
        class?: HTMLAttributes['class']
    }

    const props = withDefaults(defineProps<Props>(), {
        tag: 'p',
        class: ''
    })

    const variantClasses = computed(() => {
        const variantStyles: Record<Variant, string> = {
            h1: 'scroll-m-20 text-6xl font-medium tracking-tight',
            h2: 'scroll-m-20 text-4xl font-medium tracking-tight first:mt-0',
            h3: 'scroll-m-20 text-2xl font-medium tracking-tight',
            h4: 'scroll-m-20 text-xl font-medium tracking-tight',
            p: 'leading-6 [&:not(:first-child)]:mt-6',
            ul: 'my-6 ml-6 list-disc [&>li]:mt-2',
            li: 'leading-7',
            code: 'relative rounded bg-muted px-[0.3rem] py-[0.2rem] font-mono text-sm font-semibold'
        }

        return variantStyles[props.variant || props.tag]
    })
</script>
