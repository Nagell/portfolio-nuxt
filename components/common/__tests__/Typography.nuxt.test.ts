import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it } from 'vitest'

import Typography from '../Typography.vue'

describe('Typography Component', async () => {
    it('renders correctly', async () => {
        const wrapper = await mountSuspended(Typography, {
            slots: {
                default: 'Test Content'
            }
        })
        expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders with the correct tag', async () => {
        const text = 'Test Content'
        const wrapper = await mountSuspended(Typography, {
            props: { tag: 'h2' },
            slots: {
                default: text
            }
        })
        expect(wrapper.find('h2').exists()).toBe(true)
        expect(wrapper.find('h2').text()).toBe(text)
    })

    it('applies custom class', async () => {
        const customClass = 'custom-typography-class'
        const wrapper = await mountSuspended(Typography, {
            props: { class: customClass },
            slots: {
                default: 'Test Content'
            }
        })
        expect(wrapper.classes()).toContain(customClass)
    })
})
