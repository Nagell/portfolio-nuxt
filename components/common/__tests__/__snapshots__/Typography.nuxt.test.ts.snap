// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Typography Component > renders correctly 1`] = `"<p class="leading-6 [&amp;:not(:first-child)]:mt-6">Test Content</p>"`;
